src/DataStructures/Tensor/Expressions/TensorIndex.hpp

```diff
@@ -0,0 +1,187 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines generic tensor indices used in `TensorExpression`s
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <type_traits>
+
+#include "DataStructures/Tensor/IndexType.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace TensorIndex_detail {
+// The below values are used to separate upper indices from lower indices and
+// spatial indices from spacetime indices.
+//
+// Tensor expressions perform as many calculations as possible in a constexpr
+// context, which means working with fundamental types, specifically integer
+// types, is easiest. By using sentinel values defined in one location we can
+// easily control the encoding without having magic values floating around in
+// many places. Furthermore, encoding all the information in the `size_t` means
+// that when a failure occurs in one of the constexpr calculations it is
+// reasonably easy to debug because, while encoded, the full type information is
+// present. This approach can effectively be thought of as using specific bits
+// in the `size_t` to mark information, using the size_t more as a bitfield than
+// anything else. For human readability, we use base-10 numbers instead of
+// base-2 values that would truly set individual bits.
+//
+// Spacetime indices are represented by values [0, `spatial_sentinel`) and
+// spatial indices are represented by values
+// [`spatial_sentinel`, `max_sentinel`). Lower spacetime indices are represented
+// by values [0, `upper_sentinel`), and upper spacetime indices are represented
+// by values [`upper_sentinel`, `spatial_sentinel`). Lower spatial indices are
+// represented by values
+// [`spatial_sentinel`, `spatial_sentinel` + `upper_sentinel`), and upper
+// spatial indices are represented by values
+// [`spatial_sentinel` + `upper_sentinel`, `max_sentinel`). Values equal to or
+// above `max_sentinel` are considered invalid for representing an index.
+static constexpr size_t spatial_sentinel = 1000;
+static constexpr size_t upper_sentinel = 500;
+static constexpr size_t upper_spatial_sentinel =
+    spatial_sentinel + upper_sentinel;
+static constexpr size_t max_sentinel = 2000;
+}  // namespace TensorIndex_detail
```

Comment:
Note: I moved these sentinels into a namespace just so that they aren't global variables